README.md
# Programming 2 Group Project: The Isle of Laeso
This is our group project for Programming 2. We wanted to make a Catan style board game where you are stranded on an island with rivals and you are racing to survive and develop. 

![TheIsleOfLaeso](https://github.com/Bamboo72/Programming2GroupProject1/blob/main/res/BackgroundBoard.png)

When the game starts out every player is on their own part of the Isle of Laeso. In turn they each roll the die, move, and collect resources. Once enough resources are gathered players can build structures such as storehouses, forts, or even ports. The main goal of the game is to collect enough resources to build a port and survive the sea voyage off the island, but there are a few different ways to win. It's a race against the other players to survive the Isle of Laeso!

We decided on a viking theme to make the game interesting and distinct. From our initial research into viking mythology we learned about an island mentioned in Norse tales that is suspected to be the modern Danish island of [Læsø](https://en.wikipedia.org/wiki/L%C3%A6s%C3%B8). This was our inspiration for the name of our game.

All graphics were created in Adobe Illustrator by Jacob Schwartz, and are displayed using Swing. We decided on Swing to display our graphics becuase we had been focused on Java this school year, and decided to use what we were most used to. The programming was divided by class between the group members so that everyone had a role to play in the project development.


## Roles:
- Andrew Blodgett: Dice, Board, and Resource Programming
- Maximilian Sutton: Settings, Character Creation, and Game Turn Logic Programming
- Jacob Schwartz: Graphics, Project Management, and GUI Programming 
- Brandon Winters: Natural Events and Event Generation

## Gameplay Instructions:

After selecting the number of players and creating characters for each player, the game begins!  

In order, each player will take their turn:  
1. Roll the die using the die button.  
2. Move and collect resources by clicking on desired locations and resources. (Careful! There's no takebacks!)  
3. Choose to either build, attack, or end your turn by clicking the buttons at the bottom of the screen. Note that building or attacking will end your turn.  

There are multiple ways to win. You could escape the island, collect enough magic to ascend, or be the last player alive on the island. Keep each win condition in mind as you play!  

## Download Instructions:

Download the zip file [here](https://github.com/Bamboo72/Programming2GroupProject1/raw/main/GroupProject1.zip). Extract the contents, and run the jar file.

## Year 2 Concepts: 
- GUI
- Enhanced For Loop
- Overloading
- Overriding
- Multidimensional Arrays
- I/O and Exception Handling
- Collections
- Inheritance
- Ternary Statements

## Documentation: 
![ClassDiagram](https://github.com/Bamboo72/Programming2GroupProject1/blob/main/ConceptArt/GroupProjectClassDiagram.png)
![IslandSurvivalIdea](https://github.com/Bamboo72/Programming2GroupProject1/blob/main/ConceptArt/Isle%20of%20Laeso.jpg)



CHANGELOG.md
Added basic frame for project 2020-12-1 - Bamboo72, HoldMyTyr, andrewblodgett
 - Added Dice.java - andrewblodgett
 - Added Graphics.java - Bamboo72
 - Added Island.java - andrewblodgett
 - Added Player.java - HoldMyTyr
 - Added Recorces.java - andrewblodgett
 - Added entry point (TheIsleOfLaeso) - HoldMyTyr

  
Updated README.md 2020-12-9 - Bamboo72, HoldMyTyr, andrewblodgett

Updated Graphics 2020-12-9 - Bamboo72

Updated Dice, Recorces, and Island class 2020-12-9 - andrewblodgett

Updated Player and entry point (TheIsleOfLaeso) classes 2020-12-15 - HoldMyTyr

Updated Graphics classes 2021-1-19 - Bamboo72

Updated Graphics classes 2021-1-20 - Bamboo72

Added Consept Art 2021-1-23 - Bamboo72

Updated Recorces class 2021-1-26 - andrewblodgett

Added tempFolder w/ files 2021-1-31 - Bamboo72

Updated Dice class 2021-2-1 - andrewblodgett

Updated entry point (TheIsleOfLaeso)  2020-2-1 - HoldMyTyr

Updated Graphics 2021-2-2 - Bamboo72

Updated README.md 2021-2-2 - Bamboo72

Updated GroupProjectClassDiagram 2021-2-2 - Bamboo72

Added Graohic files 2021-2-2 - Bamboo72

Added Graohic files 2021-3-2 - Bamboo72

Updated the board classes 2021-3-6 - andrewblodgett
 - Added BoardSpace.java - andrewblodgett

  
Updated Recorces class 2021-3-6 - andrewblodgett

Added Consept Art 2021-3-8 - Bamboo72

Added Consept Art 2021-3-10 - Bamboo72

Added Consept Art  2021-3-14 - Bamboo72

Added lots of art 2021-3-15 - Bamboo72

Added Graohic files 2021-3-15 - Bamboo72

Updated IOSettings class 2021-4-5 - HoldMyTyr
 - Added IOSettings.java - HoldMyTyr
 - Added DefSetting.dat - HoldMyTyr
 - Added Setting.dat - HoldMyTyr

Updated Graphics 2021-4-6 - Bamboo72, andrewblodgett, HoldMyTyr

Updated Graphics 2021-4-7 - Bamboo72

Updated Island class 2020-7 - andrewblodgett

Updated Graphics 2021-4-8 - Bamboo72

Updated Graphics 2021-4-12 - Bamboo72

Updated IOSettings class 2021-4-13 - HoldMyTyr
 - Updated IOSettings.java - HoldMyTyr
 - Updated DefSetting.dat - HoldMyTyr
 - Updated Setting.dat - HoldMyTyr

Updated Graphics 2021-4-13 - Bamboo72

Updated Graphics 2021-4-18 - Bamboo72

Updated Player 2021-4-26 - HoldMyTyr

Updated Player 2021-4-27 - HoldMyTyr

Updated Player 2021-4-28 - HoldMyTyr

Updated Player and entry point (TheIsleOfLaeso) classes 2021-4-29 - HoldMyTyr

Updated entry point (TheIsleOfLaeso) class 2021-5-6 - HoldMyTyr

Added MousePressLissener 2021-5-9 - Bamboo72

Updated Graphics 2021-5-9 - Bamboo72

Updated Player class 2021-5-9 - Bamboo72

Updated entry point (TheIsleOfLaeso) class 2021-5-9 - Bamboo72

Updated entry point (TheIsleOfLaeso) class 2021-5-11 - HoldMyTyr

Updated Player class 2021-5-11 - HoldMyTyr

Updated Graphics 2021-5-11 - Bamboo72

Updated entry point (TheIsleOfLaeso) class 2021-5-12 - HoldMyTyr

Added Structure.java 2021-5-13 - andrewblodgett

Updated Island class 2021-5-13 - andrewblodgett

Added NaturalEventGenerator.java 2021-5-14 - bran214

Updated NaturalEventGenerator class 2021-5-14 - bran214

Updated Graphics and button classes 2021-5-16 - Bamboo72

Updated entry point (TheIsleOfLaeso) class 2021-5-18 - Bamboo72

Updated Island class 2021-5-18 - andrewblodgett

Updated MousePressLissener classes 2021-5-20 - Bamboo72

Updated Island class 2021-5-20 - andrewblodgett

Updated Graphics and Mouse classes 2021-5-20 - Bamboo72

Updated entry point (TheIsleOfLaeso) class 2021-5-20 - HoldMyTyr

Updated Structure class 2021-5-20 - andrewblodgett

Updated Island class 2021-5-20 - andrewblodgett

Updated Graphics 2021-5-20 - Bamboo72

Updated NaturalEventGenerator class 2021-5-20 - bran214

Updated Graphics 2021-5-20 - Bamboo72

Updated entry point (TheIsleOfLaeso) class 2021-5-21 - Bamboo72

Updated Structure class 2021-5-20 - bran214

Updated NaturalEventGenerator class 2021-5-21 - bran214
 - Added EnhancedNaturalEventGenerator.java - bran214

Updated entry point (TheIsleOfLaeso) class 2021-5-21 - bran214

Updated Player class 2021-5-21 - HoldMyTyr

Updated MousePressLissener classes 2021-5-20 - andrewblodgett

Updated NaturalEventGenerator class 2021-5-21 - bran214

Updated Island class 2021-5-21 - andrewblodgett

Updated entry point (TheIsleOfLaeso) class 2021-5-21 - Bamboo72

Updated entry point (TheIsleOfLaeso) class 2021-5-21 - HoldMyTyr

Updated Player class 2021-5-21 - HoldMyTyr

Updated Player class 2021-5-21 - Bamboo72

Updated Player class 2021-5-21 - HoldMyTyr

Updated MousePressLissener class 2021-5-21 - Bamboo72

Updated Player class 2021-5-21 - HoldMyTyr

Updated Player class 2021-5-21 - Bamboo72

Updated Player class 2021-5-21 - HoldMyTyr

Updated Player class 2021-5-21 - Bamboo72

Updated MousePressLissener class 2021-5-21 - Bamboo72

Updated Graphics and Player classes 2021-5-21 - Bamboo72

Update NaturalEventGenerator classes 2021-5-21 - bran214

Updated Player class 2021-5-21 - HoldMyTyr

Updated Graphics 2021-5-20 - andrewblodgett

Updated Player class 2021-5-21 - HoldMyTyr

Update NaturalEventGenerator classes 2021-5-21 - bran214

Updated Player class 2021-5-21 - HoldMyTyr

Updated IOSettings class 2021-5-21 - HoldMyTyr

Updated Player class 2021-5-21 - HoldMyTyr

Updated entry point (TheIsleOfLaeso) class 2021-5-21 - Bamboo72

Updated entry point (TheIsleOfLaeso) class 2021-5-21 - HoldMyTyr

Update NaturalEventGenerator classes 2021-5-21 - bran214

Updated Island classes 2021-5-21 - andrewblodgett

Updated entry point (TheIsleOfLaeso) class 2021-5-21 - Bamboo72

Updated Island classes 2021-5-21 - Bamboo72

Update NaturalEventGenerator classes 2021-5-21 - bran214

Updated Graphics 2021-5-20 - andrewblodgett

Updated entry point (TheIsleOfLaeso) class 2021-5-21 - HoldMyTyr

Update NaturalEventGenerator classes 2021-5-21 - bran214

Updated entry point (TheIsleOfLaeso) class 2021-5-21 - HoldMyTyr

Updated Structure classes 2021-5-21 - andrewblodgett

Updated MousePressLissener class 2021-5-21 - Bamboo72

Cleaned up Island classes 2021-5-21 - andrewblodgett

Update NaturalEventGenerator classes 2021-5-21 - bran214


src/MousePressListener.java
// Jacob Schwartz - 5/7/2021
// This is the MouseListener class for the Laeso project.
// Code copied then modified from my Maze Game project.

import java.awt.event.MouseListener;
import java.awt.event.MouseEvent;

/**
 * This class handles the mouse events that move the character
 */
class MousePressListener implements MouseListener {

    /**
     * This method detects when the mouse is clicked and moves the character
     * depending on a variety of conditions (character direction and location)
     */
    public void mouseClicked(MouseEvent event) {
        double x = event.getX();
        double y = event.getY();

        int rowNum = 0, colNum = 0;

        // System.out.println("(" + x + ", " + y + ")");

        // y /= 45.385;
        y = (y - 70) / 40;
        // y = (y - 70) / 45.385 ;
        if (0 < y && y <= 1) { // 1.54
            rowNum = 0;
        } else if (1 < y && y <= 2) {
            rowNum = 1;
        } else if (2 < y && y <= 3) {
            rowNum = 2;
        } else if (3 < y && y <= 4) {
            rowNum = 3;
        } else if (4 < y && y <= 5) {
            rowNum = 4;
        } else if (5 < y && y <= 6) {
            rowNum = 5;
        } else if (6 < y && y <= 7) {
            rowNum = 6;
        } else if (7 < y && y <= 8) {
            rowNum = 7;
        } else if (8 < y && y <= 9) {
            rowNum = 8;
        } else if (9 < y && y <= 10) {
            rowNum = 9;
        } else if (10 < y && y <= 11) {
            rowNum = 10;
        } else if (11 < y && y <= 12) {
            rowNum = 11;
        } else if (12 < y && y <= 13) {
            rowNum = 12;
        }

        // x /= 42.667;
        x = (x + 21.333) / 42.666;
        // x = (x - 21.333) / 42.667 ;
        if (0 < x && x <= 1) { // 1.54
            colNum = 0;
        } else if (1 < x && x <= 2) {
            colNum = 1;
        } else if (2 < x && x <= 3) {
            colNum = 2;
        } else if (3 < x && x <= 4) {
            colNum = 3;
        } else if (4 < x && x <= 5) {
            colNum = 4;
        } else if (5 < x && x <= 6) {
            colNum = 5;
        } else if (6 < x && x <= 7) {
            colNum = 6;
        } else if (7 < x && x <= 8) {
            colNum = 7;
        } else if (8 < x && x <= 9) {
            colNum = 8;
        } else if (9 < x && x <= 10) {
            colNum = 9;
        } else if (10 < x && x <= 11) {
            colNum = 10;
        } else if (11 < x && x <= 12) {
            colNum = 11;
        } else if (12 < x && x <= 13) {
            colNum = 12;
        } else if (13 < x && x <= 14) {
            colNum = 13;
        } else if (14 < x && x <= 15) {
            colNum = 14;
        } else if (15 < x && x <= 16) {
            colNum = 15;
        } else if (16 < x && x <= 17) {
            colNum = 16;
        } else if (17 < x && x <= 18) {
            colNum = 17;
        } else if (18 < x && x <= 19) {
            colNum = 18;
        } else if (19 < x && x <= 20) {
            colNum = 19;
        } else if (20 < x && x <= 21) {
            colNum = 20;
        } else if (21 < x && x <= 22) {
            colNum = 21;
        } else if (22 < x && x <= 23) {
            colNum = 22;
        } else if (23 < x && x <= 24) {
            colNum = 23;
        } else if (24 < x && x <= 25) {
            colNum = 24;
        } else if (25 < x && x <= 26) {
            colNum = 25;
        } else if (26 < x && x <= 27) {
            colNum = 26;
        } else if (27 < x && x <= 28) {
            colNum = 27;
        } else if (28 < x && x <= 29) {
            colNum = 28;
        } else if (29 < x && x <= 30) {
            colNum = 29;
        } else if (30 < x && x <= 31) {
            colNum = 30;
        }

        // System.out.println("Col: " + colNum + ", Row: " + rowNum);

        int currentPlayerX = TheIsleOfLaeso.players[TheIsleOfLaeso.playerTurn - 1].getXPos();
        int currentPlayerY = TheIsleOfLaeso.players[TheIsleOfLaeso.playerTurn - 1].getYPos();

        // System.out.println("Player at x: " + currentPlayerX + ", y: " +
        // currentPlayerY);

        if (TheIsleOfLaeso.g.diceRolled) { // Row = Y, Col = X, Note that I'm calling the top of the screen top/up

            if (TheIsleOfLaeso.collectsLeft > 0) { // If there are still collects left
                String theResource = TheIsleOfLaeso.i.getBoard()[rowNum][colNum];
                System.out.println("theResource: " + theResource + " so !theResource.equals(\"r \") is "
                        + !theResource.equals("r "));

                if (rowNum == currentPlayerY) { // Check the column
                    if (colNum == currentPlayerX + 1) {
                        System.out.println("To the Right");

                        if (!theResource.equals("r ")) {
                            theResource = theResource.substring(2);
                            TheIsleOfLaeso.players[TheIsleOfLaeso.playerTurn - 1].addResource(theResource);
                            // TheIsleOfLaeso.i.remove(theResource, rowNum, colNum);
                            // TheIsleOfLaeso.i.add("r ", rowNum, colNum);
                            System.out.println("On the board this is: "
                                    + TheIsleOfLaeso.i.getBoard()[rowNum][colNum].substring(2));
                            TheIsleOfLaeso.i.remove(TheIsleOfLaeso.i.getBoard()[rowNum][colNum].substring(2), colNum,
                                    rowNum);
                            TheIsleOfLaeso.collectsLeft = TheIsleOfLaeso.collectsLeft - 1;
                        }

                    }
                    if (colNum == currentPlayerX - 1) {
                        System.out.println("To the Left");
                        if (!theResource.equals("r ")) {
                            theResource = theResource.substring(2);
                            TheIsleOfLaeso.players[TheIsleOfLaeso.playerTurn - 1].addResource(theResource);
                            System.out.println("On the board this is: "
                                    + TheIsleOfLaeso.i.getBoard()[rowNum][colNum].substring(2));
                            TheIsleOfLaeso.i.remove(TheIsleOfLaeso.i.getBoard()[rowNum][colNum].substring(2), colNum,
                                    rowNum);
                            TheIsleOfLaeso.collectsLeft = TheIsleOfLaeso.collectsLeft - 1;
                        }
                    }
                }

                if (colNum == currentPlayerX) { // Check the row
                    if (rowNum == currentPlayerY + 1) {
                        System.out.println("To the Bottom");
                        if (!theResource.equals("r ")) {
                            theResource = theResource.substring(2);
                            TheIsleOfLaeso.players[TheIsleOfLaeso.playerTurn - 1].addResource(theResource);
                            System.out.println("On the board this is: "
                                    + TheIsleOfLaeso.i.getBoard()[rowNum][colNum].substring(2));
                            TheIsleOfLaeso.i.remove(TheIsleOfLaeso.i.getBoard()[rowNum][colNum].substring(2), colNum,
                                    rowNum);
                            TheIsleOfLaeso.collectsLeft = TheIsleOfLaeso.collectsLeft - 1;
                        }

                    }
                    if (rowNum == currentPlayerY - 1) {
                        System.out.println("To the Top");
                        if (!theResource.equals("r ")) {
                            theResource = theResource.substring(2);
                            TheIsleOfLaeso.players[TheIsleOfLaeso.playerTurn - 1].addResource(theResource);
                            System.out.println("On the board this is: "
                                    + TheIsleOfLaeso.i.getBoard()[rowNum][colNum].substring(2));
                            TheIsleOfLaeso.i.remove(TheIsleOfLaeso.i.getBoard()[rowNum][colNum].substring(2), colNum,
                                    rowNum);
                            TheIsleOfLaeso.collectsLeft = TheIsleOfLaeso.collectsLeft - 1;
                        }

                    }
                }

            }

            if (TheIsleOfLaeso.moveLeft > 0) { // If there are still moves left
                if (!TheIsleOfLaeso.i.getBoard()[rowNum][colNum].equals("o ")
                        && !TheIsleOfLaeso.i.getBoard()[rowNum][colNum].equals("n ")
                        && !Player.checkForPlayer(colNum, rowNum) && (Player.checkForStructure(colNum, rowNum) == null
                                || Player.checkForStructure(colNum, rowNum).owner == TheIsleOfLaeso.playerTurn)) {
                    if (colNum == currentPlayerX + 1) {
                        if (rowNum == currentPlayerY + 1) {
                            System.out.println("To the RD");
                            TheIsleOfLaeso.players[TheIsleOfLaeso.playerTurn - 1].move("ru");
                            TheIsleOfLaeso.moveLeft = TheIsleOfLaeso.moveLeft - 1;
                        }
                        if (rowNum == currentPlayerY - 1) {
                            System.out.println("To the RU");
                            TheIsleOfLaeso.players[TheIsleOfLaeso.playerTurn - 1].move("rd");
                            TheIsleOfLaeso.moveLeft = TheIsleOfLaeso.moveLeft - 1;
                        }
                    }

                    if (colNum == currentPlayerX - 1) {
                        if (rowNum == currentPlayerY + 1) {
                            System.out.println("To the LD");
                            TheIsleOfLaeso.players[TheIsleOfLaeso.playerTurn - 1].move("lu");
                            TheIsleOfLaeso.moveLeft = TheIsleOfLaeso.moveLeft - 1;
                        }
                        if (rowNum == currentPlayerY - 1) {
                            System.out.println("To the LU");
                            TheIsleOfLaeso.players[TheIsleOfLaeso.playerTurn - 1].move("ld");
                            TheIsleOfLaeso.moveLeft = TheIsleOfLaeso.moveLeft - 1;
                        }
                    }
                }

            }
            Structure theStorehouse = Player
                    .checkForPlayerAtStructure(TheIsleOfLaeso.players[TheIsleOfLaeso.playerTurn - 1], 'r');
            if (theStorehouse != null) {
                theStorehouse.emptyStorehouse();
            }
            TheIsleOfLaeso.g.boardText = TheIsleOfLaeso.players[TheIsleOfLaeso.playerTurn - 1].getName() + "'s turn. "
                    + TheIsleOfLaeso.moveLeft + " moves, " + TheIsleOfLaeso.collectsLeft + " collects left";
            TheIsleOfLaeso.g.refreshBoard();
        }
        System.out.println(TheIsleOfLaeso.i.getBoard()[rowNum][colNum]);
        System.out.println();

    }

    /**
     * Blank method
     */
    public void mouseEntered(MouseEvent event) {
    }

    /**
     * Blank method
     */
    public void mouseExited(MouseEvent event) {
    }

    /**
     * Blank method
     */
    public void mousePressed(MouseEvent event) {
    }

    /**
     * Blank method
     */
    public void mouseReleased(MouseEvent event) {
    }

}